File: backend/src/llm/promptTemplate.js
/**
 * 将提示词模板中的 ${变量名} 替换为用户资料中的实际内容。
 * 支持的变量：个人信息、人生战略、个人记忆
 */
const VARIABLE_RESOLVERS = {
  个人信息: (user) => user.info ?? '',
  人生战略: (user) => user.lifeStrategy ?? '',
  个人记忆: (user) => user.personalMemory ?? '',
};

/** 解析并拼接最终系统提示词 */
export function resolvePromptTemplate(template, user) {
  if (!template || typeof template !== 'string') return '';

  const resolved = template.replace(/\$\{([^}]+)\}/g, (_, rawKey) => {
    const key = rawKey.trim();
    const resolver = VARIABLE_RESOLVERS[key];
    return resolver ? resolver(user) : '';
  });

  return resolved.trim();
}

/**
 * 组装发给大模型的消息列表：系统提示词 + 对话历史。
 * @param {Array<{ sender: string, text: string }>} thread 存储层的消息线程
 * @param {object} user 用户资料
 * @param {function} toLlmMessages 将线程转为 LLM 角色的函数
 */
export function buildLlmMessages(thread, user, toLlmMessages) {
  const history = toLlmMessages(thread);
  const systemContent = resolvePromptTemplate(user.promptTemplate, user);

  if (!systemContent) return history;

  return [{ role: 'system', content: systemContent }, ...history];
}

File: backend/app/llm/prompt_template.py
"""提示词模板解析。"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

VARIABLE_RESOLVERS = {
    "个人信息": lambda user: user.get("info") or "",
    "人生战略": lambda user: user.get("lifeStrategy") or "",
    "个人记忆": lambda user: user.get("personalMemory") or "",
}


def resolve_prompt_template(template: Optional[str], user: Dict[str, Any]) -> str:
    if not template or not isinstance(template, str):
        return ""

    def replace(match: Any) -> str:
        key = match.group(1).strip()
        resolver = VARIABLE_RESOLVERS.get(key)
        return resolver(user) if resolver else ""

    import re

    resolved = re.sub(r"\$\{([^}]+)\}", replace, template)
    return resolved.strip()


def to_llm_messages(thread: List[Dict[str, Any]]) -> List[Dict[str, str]]:
    return [
        {
            "role": "user" if message.get("sender") == "me" else "assistant",
            "content": message.get("text", ""),
        }
        for message in thread
    ]


def build_llm_messages(
    thread: List[Dict[str, Any]],
    user: Dict[str, Any],
) -> List[Dict[str, str]]:
    history = to_llm_messages(thread)
    system_content = resolve_prompt_template(user.get("promptTemplate"), user)
    if not system_content:
        return history
    return [{"role": "system", "content": system_content}, *history]
